halo2-base/src/poseidon/hasher/state.rs

```diff
@@ -23,7 +26,73 @@ impl<F: ScalarField, const T: usize, const RATE: usize> PoseidonState<F, T, RATE
         Self { s: default_state.map(|f| ctx.load_constant(f)) }
     }
 
-    pub fn x_power5_with_constant(
+    /// Perform permutation on this state.
+    ///
+    /// `inputs` is right padded.
+    /// If `len` is `None`, treat `inputs` as a fixed length array.
+    pub fn permutation(
+        &mut self,
+        ctx: &mut Context<F>,
+        gate: &impl GateInstructions<F>,
+        inputs: &Vec<AssignedValue<F>>,
+        len: Option<AssignedValue<F>>,
+        spec: &OptimizedPoseidonSpec<F, T, RATE>,
+    ) {
+        let r_f = spec.r_f / 2;
+        let mds = &spec.mds_matrices.mds.0;
+        let pre_sparse_mds = &spec.mds_matrices.pre_sparse_mds.0;
+        let sparse_matrices = &spec.mds_matrices.sparse_matrices;
+
+        // First half of the full round
+        let constants = &spec.constants.start;
+        if let Some(len) = len {
+            // Note: this doesn't mean `padded_inputs` is 0 padded because there is no constraints on `inputs[len..]`
+            let padded_inputs: [AssignedValue<F>; RATE] =
+                core::array::from_fn(
+                    |i| if i < inputs.len() { inputs[i] } else { ctx.load_zero() },
```

Comment:
rust trick: `inputs.get(i).unwrap_or_else(|| ctx.load_zero())`

Reply:
I tried this but met a problem - `get(i)` returns a reference but `ctx.load_zero()` is a temporarily value which cannot be ref.